core/modules.rs

```diff
@@ -45,135 +48,185 @@ pub trait Loader {
-// combined into one type.
-pub enum Either<E> {
-  JSError(JSError),
-  Other(E),
+struct PendingLoad<E: Error> {
+  url: String,
+  is_root: bool,
+  source_code_future: Box<SourceCodeFuture<E>>,
 }
 
 /// This future is used to implement parallel async module loading without
-/// complicating the Isolate API.
-pub struct RecursiveLoad<'l, L: Loader> {
-  loader: &'l mut L,
-  pending: HashMap<String, Box<SourceCodeFuture<<L as Loader>::Error>>>,
-  root: String,
+/// complicating the Isolate API. Note that RecursiveLoad will take ownership of
+/// an Isolate during load.
+pub struct RecursiveLoad<L: Loader> {
+  loader: Option<L>,
+  pending: Vec<PendingLoad<L::Error>>,
+  is_pending: HashSet<String>,
+  phantom: PhantomData<L>,
+  // TODO(ry) The following can all be combined into a single enum State type.
+  root: Option<String>,           // Empty before polled.
+  root_specifier: Option<String>, // Empty after first poll
+  root_id: Option<deno_mod>,
 }
 
-impl<'l, L: Loader> RecursiveLoad<'l, L> {
+impl<L: Loader> RecursiveLoad<L> {
   /// Starts a new parallel load of the given URL.
-  pub fn new(url: &str, loader: &'l mut L) -> Self {
-    let root = L::resolve(url, ".");
-    let mut recursive_load = Self {
-      loader,
-      root: root.clone(),
-      pending: HashMap::new(),
+  pub fn new(url: &str, loader: L) -> Self {
+    Self {
+      loader: Some(loader),
+      root: None,
+      root_specifier: Some(url.to_string()),
+      root_id: None,
+      pending: Vec::new(),
+      is_pending: HashSet::new(),
+      phantom: PhantomData,
+    }
+  }
+
+  fn take_loader(&mut self) -> L {
+    self.loader.take().unwrap()
+  }
+
+  fn add(
+    &mut self,
+    specifier: &str,
+    referrer: &str,
+    parent_id: Option<deno_mod>,
+  ) -> Result<String, L::Error> {
+    let url = L::resolve(specifier, referrer)?;
+
+    let is_root = if let Some(parent_id) = parent_id {
+      let loader = self.loader.as_mut().unwrap();
+      let modules = loader.modules();
+      modules.add_child(parent_id, &url);
+      false
+    } else {
+      true
     };
-    recursive_load
-      .pending
-      .insert(root.clone(), recursive_load.loader.load(&root));
-    recursive_load
+
+    if !self.is_pending.contains(&url) {
+      self.is_pending.insert(url.clone());
+      let source_code_future = {
+        let loader = self.loader.as_mut().unwrap();
+        loader.load(&url)
+      };
+      self.pending.push(PendingLoad {
+        url: url.clone(),
+        source_code_future,
+        is_root,
+      });
+    }
+
+    Ok(url)
   }
 }
 
-impl<'l, L: Loader> Future for RecursiveLoad<'l, L> {
-  type Item = deno_mod;
-  type Error = Either<L::Error>;
+// TODO(ry) This is basically the same thing as RustOrJsError. They should be
+// combined into one type.
+#[derive(Debug, PartialEq)]
+pub enum Either<E> {
+  JSError(JSError),
+  Other(E),
+}
+
+impl<L: Loader> Future for RecursiveLoad<L> {
+  type Item = (deno_mod, L);
```

Comment:
I find that returning the loader from the future is really ugly and reeks. In the future we should try to refactor this.

Reply:
refactors welcome